### DISCUSSION TITLE
Introduce ltsp-staging

### DISCUSSION THREAD

**Discussion by alkisg** [ltsp/ltsp#212] [Q&A]
At times, some developers propose implementations for useful things, but for some reasons they shouldn't be in the default installation yet. Those reasons might include lack of time for peer review, limited audience, to allow time for the ideas and their implementations to mature etc.

We could introduce a new https://github.com/ltsp/staging repository where these implementations would live. A recipe would build this and produce different and optional packages in the PPA:

- An ltsp-staging package would contain all the non-intrucive code; i.e. it would mainly add files and applets in /usr/share/ltsp that administrators would optionally use.
- Possibly other packages like for example `ltsp-rbd-initramfs`, which, when installed, would add RBD (or HTTP or AoE) support booting in initramfs-tools.

With this design, ltsp.org, Debian, Ubuntu etc wouldn't know anything about staging, but it would be available in the PPA, supported in ltsp and community issues etc. And later on, some of those features would migrate into the main repository.

Note the restriction here: the changes to the main ltsp code should be absolutely minimal. LTSP is modular, but it can't do everything, some new features can't be implemented as "plugins" in staging.

With that restriction, are there any developers that would be interested in me setting up staging for them? @kvaps? @trickkiste? If changes to the main ltsp code are always necessary, then staging isn't appropriate, and we'd then want to provide an infrastructure for "feature-forks" instead of "staging".

**Comment by kvaps** [comment]
Hi @alkisg I like this idea, but how are you going to set up the development?
How and when do you want make releases and how to synchronize it with the main code base?
Who will maintain this project?

> Note the restriction here: the changes to the main ltsp code should be absolutely minimal. LTSP is modular, but it can't do everything, some new features can't be implemented as "plugins" in staging.

As my usecase (building network-bootable server farm without GUI for the kubernetes) is little different than yours (building linux terminal server for users). My idea was to split LTSP to the editions, eg I can maintain the `ltsp-cloud` or even `lcsp` (Linux Cloud Server Project) - special edition for the cloud usage. Which can have own defaults, fave extra features (eg. docker support, HTTP-booting for the rootfs) and have no some other things from the main code base (eg. some cleanup procedures and ltsp pam module).

**Comment by alkisg** [comment]
The problems I'm trying to solve are:

- I won't have time for LTSP for a year; then I'll again have a lot of time for it.
- I want the LTSP 5 to 19 upgrade to be as smooth as possible for existing LTSP users.
- Which means that everything should be tested a lot for the most frequent use cases. I have done that part for the current code base, so we're pretty much ready for Ubuntu 20.04.
- New developers like you or Markus, want to see a lot of new things in LTSP and are willing to work for that, and that's awesome, but since I don't have time to test, it might break things for the other use cases that you don't use.
- I do not want to block progress.

So my proposal essentially means "let's find a way for you to submit code, have a PPA and users testing it, get feedback, let it mature etc, and when I have time I'll properly examine all forks/patches/etc and integrate in the main LTSP whatever is best for most LTSP users, and possibly separate the rest into other packages or plugins".

So staging is a temporary thing, even if it means "a year" in this particular case; after that, the staging code should either join LTSP or become separate packages/plugins (with LTSP providing appropriate hooks for them).

**Comment by alkisg** [comment]
Considering https://github.com/ltsp/ltsp/pull/90#issuecomment-557795573 as well, staging might indeed be the same as a "temporary fork" for now; and we could examine next year how many of the new features could be merged and what changes would need to be made both in the features and in the main ltsp.

**Comment by kvaps** [comment]
> So my proposal essentially means "let's find a way for you to submit code, have a PPA and users testing it, get feedback, let it mature etc, and when I have time I'll properly examine all forks/patches/etc and integrate in the main LTSP whatever is best for most LTSP users, and possibly separate the rest into other packages or plugins".

I'm not arguing I like this idea, let's do that.

What about the testing, well we can automate the testing using guestfish or some other tool in the future. Any way, I'm checking all my changes because I'm going to use them for myself and going to present them in our community.

>  So staging is a temporary thing, even if it means "a year" in this particular case; after that, the staging code should either join LTSP or become separate packages/plugins (with LTSP providing appropriate hooks for them).

Why not simple use `staging` branch for existing project and keep stable `master` frozen for the Ubuntu 20.04 integration?

**Comment by alkisg** [comment]
> I'm not arguing I like this idea, let's do that.

Done; you should now have maintainer access to https://github.com/ltsp/staging; one way to proceed could be that you push a kvaps branch there, which will be built into a PPA (will do that when we decide what we want).

If your efforts result in a different package, e.g. ltsp-cloud, it may build in the same PPAs as LTSP.
If they result in a "different version of LTSP, that could be completely integrated next year", i.e. if you'll be using the "ltsp" package name, they should go in a different PPA, e.g. we can make an ltsp-kvaps PPA.

> What about the testing, well we can automate the testing using guestfish or some other tool in the future

Sure, we can do some things, but unfortunately netbooting varies so much that the best testing can only happen manually. E.g. you can't know that this XXX realtek driver will fail and this YYY intel driver will succeed in netbooting via undionly.kpxe.

>  I'm checking all my changes because I'm going to use them for myself and going to present them in our community.

Oh I'm sure you're checking your commits, but it's not easy to know that e.g. ${next-server} in ltsp-dnsmasq.conf won't work in proxyDHCP, if you're not even using proxyDHCP. That's why I want to make sure that changes do not affect negatively the existing LTSP use cases (when I do have the time for it).

> Why not simple use staging branch for existing project and keep stable master frozen for the Ubuntu 20.04 integration?

trickkiste might want to implement an RBD branch which may conflict with yours; kimheino might implement a dracut branch with no impact on the Ubuntu code that I might want to integrate immediately; etc. Since I won't have time for peer review for a year, I won't even be able to coordinate new contributors much.

Thanks!

**Comment by kvaps** [comment]
>   Done; you should now have maintainer access to https://github.com/ltsp/staging; one way to proceed could be that you push a kvaps branch there, which will be built into a PPA (will do that when we decide what we want).

Great 🎉 I'd like to develop `ltsp-cloud` branch and eponymous package as part of the main PPA. This will allow us to maintain shared packages (like `ltsp-binaries`) together and still have different LTSP-versions for each purpose.
Users could simple switch between the versions, the same way like choosing the kernel:
```
linux-image-generic
  Generic Linux kernel image

linux-image-aws
  Linux kernel image for Amazon Web Services (AWS) systems.

linux-image-azure
  Linux kernel image for Azure systems.

linux-image-virtual
  Virtual Linux kernel image
...
```

User could choose the needed LTSP version for his purpose:

```
ltsp
  Linux Terminal Server Project

ltsp-cloud
  Linux Terminal Server Project (enhanced with cloud features)

ltsp-rbd
  Linux Terminal Server Project (enhanced for RBD support)
...
```

I'd like to not change the code much to leave the opportunity to merge these projects into one in the distant future. But I'd like to change some defaults (eg. predefine some setting in ltsp.conf file)

I think all these changes, including the package name, can be described as single patch for the main code base.

BTW, what about Pull Request policy, should the be done in main project, on staging or on both?

**Comment by alkisg** [comment]
That sounds great. For now, please limit pull requests to minimal features (e.g. the http support in ltsp-dnsmasq.conf, but not the BOOT_METHOD change) and file them against the ltsp branch; next year, I'll have a look at all the nice things you'll have done, see which of them best fit the ltsp branch etc, and we'll organize the merge then.

When you're ready, ping me to prepare a PPA recipe for ltsp-cloud.

**Comment by kvaps** [comment]
Ok, I've prepared ltsp-cloud branch and [merged](https://github.com/ltsp/staging/commits/ltsp-cloud) all my PRs into it (except ltsp/ltsp#83),

I can prepare ltsp-cloud package:

```bash
mv debian/ltsp.dirs debian/ltsp-cloud.dirs
mv debian/ltsp.install debian/ltsp-cloud.install
mv debian/ltsp.links debian/ltsp-cloud.links
mv debian/ltsp.manpages debian/ltsp-cloud.manpages
mv debian/ltsp.triggers debian/ltsp-cloud.triggers

sed -i debian/changelog \
    -e 's|^ltsp |ltsp-cloud |g'

sed -i debian/rules \
    -e 's|/debian/ltsp/|/debian/ltsp-cloud/|'

sed -i debian/control \
    -e 's|\(https://github.com/ltsp/\)ltsp|\1staging|' \
    -e 's/^\(Source\|Package\): .*/\1: ltsp-cloud/' \
    -e '/^\(Replaces\|Conflicts\): /d' \
    -e '/^Architecture: /iReplaces: ltsp' \
    -e '/^Architecture: /iConflicts: ltsp'

dpkg-buildpackage -b -rfakeroot -us -uc
```

What next?

**Comment by alkisg** [comment]
First of all go ahead and push those debian/* changes regarding ltsp-cloud.
Then we need to get this branch to launchpad, in order to automatically build packages from it.

I could set up a [code import from github](https://code.launchpad.net/+code-imports/+new), but that has issues when people force-push in github.

Let's do it this way for now: modify your .git/config so that it pushes in two remotes at the same time:

```shell
[remote "origin"]
        url = git@github.com:ltsp/staging.git
        fetch = +refs/heads/*:refs/remotes/origin/*
        pushurl = git@github.com:ltsp/staging.git
        pushurl = git+ssh://kvaps@git.launchpad.net/~kvaps/ltsp/+git/staging
```

Then I'll make a recipe that will use the ltsp-cloud branch of your ltsp/staging launchpad git repository, and daily build the result in the PPA.

**Comment by kvaps** [comment]
> First of all go ahead and push those debian/* changes regarding ltsp-cloud.

That's done, but I little confused about versioning, I was thinking to follow semver and stick versioning with main package. But semver is not describes upstream feature forks case, it [describes](https://semver.org/#spec-item-11) only pre-releases. So `19.11-cloud.1-1` will considered always **before** `19.11-1`. I may use additional patch version, (eg. `19.11.1-1`, `19.11.2-1`...) or use package version (eg. `19.11-2`, `19.11-3`...)
Could you give me an advice, please?

> I could set up a code import from github, but that has issues when people force-push in github.

That's shouldn't be a problem we can protect the branch from force-pushes, this is always bad practices

**Comment by alkisg** [comment]
We're using calver: https://ltsp.org/advanced/versioning/

You don't need to modify the versioning at all. You have a different package name that should Conflicts: ltsp. The version can be the same, e.g. 19.11-1\~201911110854\~ubuntu18.04.1.

Re: force pushes, OK, then you don't need to do anything in launchpad or in your git setup, I'll set up an import.

**Comment by kvaps** [comment]
> You don't need to modify the versioning at all. You have a different package name that should Conflicts: ltsp. The version can be the same, e.g. 19.11-1~201911110854~ubuntu18.04.1.

Shouldn't I describe changes in changelog for each release?

**Comment by alkisg** [comment]
It's up to you. It's good practice to do that. But I don't think you need a history of releases. So, one single changelog entry that you would update each time with the current date as the release.

Example:

```
ltsp (20.01-1) unstable; urgency=low

  * Implement this (#xx)
  * Fix that (#yy)

 -- kvaps ...  date

ltsp (19.11-1) unstable; urgency=low

  * Prefer external ltsp-binaries package, fall back to ipxe (#49)
...
 -- Alkis Georgopoulos <....>  Mon, 11 Nov 2019 09:52:19 +0200
```

And if you ever need to rebase with a newer LTSP release, you can import the upstream changelog and again only keep your one additional entry.

**Comment by kvaps** [comment]
OK, I've already prepared first release:
https://github.com/ltsp/staging/commit/34df791dbd9527b36a808d7401215dea19ebd430

> And if you ever need to rebase with a newer LTSP release, you can import the upstream changelog and again only keep your one additional entry.

That's true. I think that using calver is good choose for having loosely coupled version with the upstream package, and still understandable for comparing perform rebase operations.
So, I will use own versioning based on calver detached from upstream.

> Re: force pushes, OK, then you don't need to do anything in launchpad or in your git setup, I'll set up an import.

Could you also protect the `ltsp-cloud` branch, I have no permissions for doing that.

**Comment by alkisg** [comment]
I've prepared the [launchpad imported branch](https://code.launchpad.net/~ltsp/ltsp/+git/staging) and an [ltsp-cloud recipe](https://code.launchpad.net/~ltsp/+recipe/ltsp-cloud) and I've configured daily builds to the LTSP proposed PPA. Ping me if anything is amiss.

I think a wiki page would be nice too (or pointer to your own site/blog/etc), and maybe a community issue to notify the users and enable discussions about ltsp-cloud.

**Comment by alkisg** [comment]
Builds failed: https://code.launchpad.net/~ltsp/+recipe/ltsp-cloud

You have one less space in line 13 in debian/changelog, two spaces are needed before the date. Not sure if it was caused because of that one.

**Comment by kvaps** [comment]
I updated the double space, could you restart the build?

**Comment by alkisg** [comment]
Sure, done.

**Comment by kvaps** [comment]
Well, as can I see it is still uses https://github.com/ltsp/staging/commit/34df791dbd9527b36a808d7401215dea19ebd430 with single space instead of https://github.com/ltsp/staging/commit/5d5a0b354320f820c56c2fe73b6e167363cb68f2 (which is current HEAD), that's probably problem what you talking about

**Comment by alkisg** [comment]
No this was because it imports automatically after 5 hours (while pushing with dual remote origins is instant).
I clicked on "import now" and requested another build, let's see.

**Comment by kvaps** [comment]
OK, then I'd like to have access for push to lp for save your time next time, could you provide it to me too?

**Comment by alkisg** [comment]
The lp branch for the recipe can be a personal one, like I wrote [above](https://github.com/ltsp/ltsp/issues/93#issuecomment-557995031):

```
        pushurl = git+ssh://kvaps@git.launchpad.net/~kvaps/ltsp/+git/staging
```

(assuming kvaps is your launchpad username). Launchpad recipes can use any launchpad git tree as a source.

**Comment by kvaps** [comment]
All right! Yes, it's mine. I've just successfully pushed to the launchpad.
Thanks for help!

**Comment by alkisg** [comment]
I deleted the old recipe and created a new one on the same URL, that uses your launchpad branch, and invoked a build:

https://code.launchpad.net/~ltsp/+recipe/ltsp-cloud

**Comment by alkisg** [comment]
That last build failed due to force-pushing to an older date:
https://code.launchpad.net/~kvaps/ltsp/+git/staging/+ref/ltsp-cloud
has different history from:
https://code.launchpad.net/~ltsp/ltsp/+git/staging/+ref/ltsp-cloud

I think it'll be OK if you push again with a more current date. I.e. you can now force push, but it can't build older dates over new dates. Example on how to amend a date:

```
GIT_COMMITTER_DATE="Wed Feb 16 14:00 2011 +0100" git commit --amend
```

**Comment by kvaps** [comment]
Got it!

**Comment by trickkiste** [comment]
Hi Alkis & Co,

due to illness, I did not get around to appreciate this move yet.

Everything, which allows for progress and experiments is
greatly appreciated of course. Some thoughts I have on the matter though.

Seems like the "staging" idea has already been abandoned, or more correctly
been replaced with the idea of "named branches". That actually might be the
best way to go anyways, as it allows for self-organization. I would hope
this would turn out in such a way, that over time we will see one or two
well maintained "personal" branches, which already incorporate the best
features submitted for inclusion - somewhat like the late -mm kernel tree
of Andrew Morton, today known as "linux-next".

Having just a bunch of branches flying around, I believe, is not ideal. An
"alternative version" does need a maintainer, just like the stable version.
Also, I think such a version would have to be named "testing" or
"experimental" not "staging", as staging usually is seen more as a release
candidate, a version getting ready for release, already containing only the
features selected for inclusion in the next stable version. However,
"personal" branches are even better, as we don't have to designate in
advance, which person maintains the "testing" branch. One of the personal
branches would emerge as the defacto "ltsp-next" branch due to the forces
of self-organisation.

Also, if we go down that road, I see no point in restricting patches to
non-core changes. There is no need for that. If this openness leads to a
defacto "ltsp-next" branch, it's maintainer should be able to select
improvements of any kind for his testing version. And if your final tests,
Alkis, prove the testing version worthy, it would be promoted to staging
and then stable.

Openness prevails, self organisation rulez.

My best regards,
Markus

PS: Some facilities for automated building are greatly appreciated, docs on
the build tools currently used would be very helpful too.

Am Mo., 25. Nov. 2019 um 17:33 Uhr schrieb kvaps <notifications@github.com>:

**Comment by alkisg** [comment]
The main idea of staging was "not enough time currently to properly review and integrate".
Spending more time on setting up a better staging infrastructure contradicts the main reason it exists.

I do not expect staging to exist after a couple of years, as then I'll have more time for LTSP, and we'll have more developers familiar with the code base and the existing LTSP user needs, and more code maturity, making staging or named branches unnecessary.

Until then, I'm moving this to the community issues, in case Markus or anyone else needs help in setting up their temporary ltsp feature branches.
